docs/2019-03-28-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,253 @@
+the two are unifiable.
+
+Formally, the algorithm simplifies `φ(x) ∧ ¬∃y.ψ(x,y)` to `φ(x) ∧ ¬∃y.⌈φ(x) ∧ ψ(x,y)⌉`
+
+### Applying circularities
+
+To apply circularities we have to have made progress using regular rules.
+However, whenever circularities may be applied, we want to apply them
+immediately and to only apply regular rules on the parts on which
+circularities cannot apply.
+
+Therefore, if we are not at the first step in proving the claim,
+we want to attempt applying all claims, similarly to how
+we do that for the one-path reachability claims in the corresponding algorithm.
+
+Doing so, we will obtain one result (maybe `⊥`) for each of the applied claims
+and the remainder part, i.e., the part on which none of the claims
+could be applied:
+
+```
+φ'ᵣₑₘ := φᵣₑₘ ∧ ¬∃x₁.⌈φ∧φ₁⌉ ∧ … ∧ ¬∃xₙ.⌈φ∧φₙ⌉ 
+```
+
+We have a chioce whether to apply circularities sequentially or in parallel.
+
+#### Applying claims sequentially
+
+```
+∀x.φ → [w]∃z.ψ                                              is equivalent to
+∀x.φ ∧ (∃xᵢ.φᵢ ∨ ¬∃xᵢ.φᵢ) → [w]∃z.ψ                         which is equivalent to
+∀x.(φ ∧ ∃xᵢ.φᵢ) ∨ (φ ∧ ¬∃xᵢ.φᵢ) → [w]∃z.ψ                   which is equivalent to
+∀x.((φ ∧ ∃xᵢ.φᵢ) → [w]∃z.ψ) ∧ ∀x. ((φ ∧ ¬∃xᵢ.φᵢ) → [w]∃z.ψ) (1)
+```
+
+Note that the remainder `∀x.φ ∧ ¬∃xᵢ.φᵢ → [w]∃z.ψ` can be rewritten as
+`∀x.φ ∧ ¬∃xᵢ.⌈φ∧φᵢ⌉ → [w]ψ`, as detailed above.
+
+__Note:__ If there are multiple claims which could apply on the same concrete
+instance of a configuration, then applying them sequentially would reduce
+the search space by potentially creating less goals.
+
+If so, then `φ` for claim `i+1` would be
+`φᵢ₊₁ʳᵉᵐ := φ ∧ ¬∃z.⌈φ∧ψ⌉ ∧ ¬∃x₁.⌈φ∧φ₁⌉ ∧ …  ∧ ¬∃xᵢ.⌈φ∧φᵢ⌉`
+
+#### Applying claims in parallel
+
+Note that in the above resoning we have used the idea that the remainder
+is computed at every claim application; it is equally possible
+to do it only once for all claims:
+
+```
+∀x.φ → [w]∃z.ψ
+∀x.φ ∧ (∃x₁.φ₁ ∨ … ∨ ∃x₁.φₙ ∨ (¬∃x₁.φ₁ ∧ … ∧ ¬∃xₙ.φₙ) → [w]∃z.ψ  
+∀x.(φ ∧ ∃x₁.φ₁) ∨ … ∨ (φ' ∧ ∃xₙ.φₙ) ∨ (φ' ∧ (¬∃x₁.φ₁ ∧ … ∧ ¬∃xₙ.φₙ)) → [w]∃z.ψ
+(∀x.φ ∧ ∃x₁.φ₁ → [w]∃z.ψ)  ∧ … (∀x.φ ∧ ∃xₙ.φₙ → [w]∃z.ψ) ∧ (∀x.(φ ∧ (¬∃x₁.φ₁ ∧ … ∧ ¬∃xₙ.φₙ)) → [w]∃z.ψ)
+```
+
+Note that the remainder can be rewritten as:
+
+```
+φ ∧ (¬∃x₁.φ₁ ∧ … ∧ ¬∃xₙ.φₙ) → [w]ψ
+(φ ∧ ¬∃x₁.φ₁) ∧ … ∧ (φ ∧ ¬∃xₙ.φₙ) → [w]ψ
+(φ ∧ ¬∃x₁.⌈φ ∧ φ₁⌉) ∧ … ∧ (φ ∧ ¬∃xₙ.⌈φ∧φₙ⌉) → [w]ψ
+φ ∧ (¬∃x₁.⌈φ ∧ φ₁⌉ ∧ … ∧  ¬∃xₙ.⌈φ∧φₙ⌉) → [w]ψ
+```
+
+The advantage of this approach is that it's simpler, not altering the starting goal
+from one derivation to the next.
+
+#### Using a claim to advance the corresponding goal
+
+We want to prove that from
+```
+∀xᵢ.φᵢ → [w]∃zᵢ.ψᵢ
+∀x∪zᵢ.ψᵢ ∧ ∃xᵢ.⌈φ ∧ φᵢ⌉ → [w]ψ
+```
+we can deduce that `∀x.φ ∧ ∃xᵢ.φᵢ → [w]∃z.ψ`.
+
+This would allow us to replace goal `∀x.φ ∧ ∃xᵢ.φᵢ → [w]∃z.ψ`
+with goal '∀x∪zᵢ.ψᵢ ∧ ∃xᵢ.⌈φ ∧ φᵢ⌉ → [w]ψ'.
+
+### Applying axioms
+
+We're back now to `∀x.φ →  (∃z.○[w]ψ) ∧ •⊤`, which is equivalent to 
+`(∀x.φ →  ○[w]ψ) ∧ (∀x.φ →  •⊤)`
```

Comment:
why ∃z is omitted here?

Reply:
by mistake.  thanks for spotting it.

Reply:
I did not have `∃z` initially, but then I realized we need them, as they introduce variables which become universally quantified in the derived goals.  I tried to refactor and add them everywhere, but it seems I missed some spots.